### PR TITLE
SRE-3 add opentelemetry to pbnj

### PR DESCRIPTION
**Pull request by tobert**
## Description

Adds OpenTelemetry plumbing and basic traces to PB&J.

## Why is this needed

Distributed tracing.

## How Has This Been Tested?

go test ./... & CI

## How are existing users impacted? What migration steps/scripts do we need?

PowerSet now always calls out to get the current power status whether it needs to or not.

^^ is reported in traces along with the desired power action.

With the OTEL_EXPORTER_OTLP_ENDPOINT set to a non-empty value, PB&J will attempt to connect to it and send traces. When it's not set, traces will be in non-recording mode and there is no change in behavior.

### CODE REVIEW

File: grpc/oob/machine/machine.go [hunk 1]
```diff
@@ -225,24 +264,29 @@ func (m Action) PowerSet(ctx context.Context, action string) (result string, err
...
+	span.SetAttributes(attribute.String("currentPowerState", currentPowerState))
+
 	ok := true
 	if pwrAction == "status" {
-		result, err = client.GetPowerState(ctx)
+		result = currentPowerState
 	} else {
 		if action == v1.PowerAction_POWER_ACTION_CYCLE.String() {
 			// check status
 			// if powered on, do cycle
```
Comment (nshalman) [inline grpc/oob/machine/machine.go:287]: It seems like this part of the comment may not have been accurate and remains inaccurate... Should it have been? Should we make it accurate (again?) ? CC @jacobweinstock

**Review by nshalman — COMMENTED** [review]
This looks good to me, but I'd like Jacob to review it too.

**Comment by nshalman** [comment]
I was able to test this locally with Jaeger and confirm that it works and can still do things:
Screen Shot 2021-11-12 at 4 28 54 PM

**Review by nshalman — APPROVED** [review]

File: grpc/oob/machine/machine.go [hunk 2]
```diff
@@ -225,24 +264,29 @@ func (m Action) PowerSet(ctx context.Context, action string) (result string, err
 	log.Info("connected to BMC", logMetadata(client.GetMetadata())...)
 	m.SendStatusMessage("connected to BMC")
 
+	// fetch the current power state that will be returned by "status"
+	// or used for control in cycle, and is always sent in traces
+	currentPowerState, err := client.GetPowerState(ctx)
+	if err != nil {
+		log.Error(err, "failed to get power state")
```
Comment (nshalman) [inline grpc/oob/machine/machine.go:271]: It looks like we're missing a span.SetStatus here where this error is being logged.

File: grpc/oob/machine/machine.go [hunk 3]
```diff
@@ -251,14 +295,17 @@ func (m Action) PowerSet(ctx context.Context, action string) (result string, err
 	}
 	log = m.Log.WithValues(logMetadata(client.GetMetadata())...)
```
Comment (nshalman) [inline grpc/oob/machine/machine.go:296]: The data from client.GetMetadata() would be very valuable for annotating the trace.

File: grpc/oob/machine/machine.go [hunk 4]
```diff
@@ -116,16 +133,20 @@ func (m Action) BootDeviceSet(ctx context.Context, device string, persistent, ef
...
 	}
+	span.SetAttributes(attribute.String("bmc.host", host), attribute.String("bmc.username", user))
 	base := "setting boot device: " + m.BootDeviceRequest.GetBootDevice().String()
 	msg := "working on " + base
 	m.SendStatusMessage(msg)
+
 	client := bmclib.NewClient(host, "623", user, password, bmclib.WithLogger(m.Log))
 
 	m.SendStatusMessage("connecting to BMC")
 	err = client.Open(ctx)
```
Comment (nshalman) [inline grpc/oob/machine/machine.go:147]: After this call to Open, the client metadata would be very helpful to annotate the span.

File: grpc/oob/machine/machine.go [hunk 5]
```diff
@@ -225,24 +264,29 @@ func (m Action) PowerSet(ctx context.Context, action string) (result string, err
 	log.Info("connected to BMC", logMetadata(client.GetMetadata())...)
 	m.SendStatusMessage("connected to BMC")
 
+	// fetch the current power state that will be returned by "status"
+	// or used for control in cycle, and is always sent in traces
+	currentPowerState, err := client.GetPowerState(ctx)
```
Comment (nshalman) [inline grpc/oob/machine/machine.go:269]: After the call to GetPowerState (and also to SetPowerState elsewhere) the client metadata would be valuable to annotate the span

**Review by jacobweinstock — APPROVED** [review]

**Pull request merged**